lib/zhr_devs/submissions/commands/submit_solution.ex

```diff
@@ -0,0 +1,100 @@
+defmodule ZhrDevs.Submissions.Commands.SubmitSolution do
+  @moduledoc """
+  A command must contain a field to uniquely identify the aggregate instance (e.g. account_number).
+  Use @enforce_keys to force the identity field to be specified when creating the command struct.
+  """
+  @fields [:uuid, :hashed_identity, :technology, :task_uuid, :solution_path, :submission_identity]
+  @enforce_keys @fields
+  defstruct @fields
+
+  use Witchcraft.Comonad
+
+  alias Uptight.Result
+  alias Uptight.Text, as: T
+
+  alias ZhrDevs.App
+
+  alias ZhrDevs.Submissions.SubmissionIdentity
+
+  import Uptight.Assertions
+
+  @supported_technologies Enum.map(
+                            Application.compile_env(:zhr_devs, :supported_technologies),
+                            &Atom.to_string/1
+                          )
+
+  @typep error() :: String.t() | struct()
+
+  @spec dispatch(Keyword.t()) :: :ok | {:error, error()}
+  def dispatch(opts) do
+    case parse(opts) do
+      %Uptight.Result.Ok{} = ok_result ->
+        ok_result
+        |> Result.from_ok()
+        |> App.dispatch()
+
+      error ->
+        {:error, error |> extract() |> build_error()}
+    end
+  end
+
+  #### Private functions ####
+
+  defp parse(opts) do
+    Result.new(fn ->
+      hashed_identity =
+        opts
+        |> Keyword.fetch!(:hashed_identity)
+        |> Uptight.Base.mk_url!()
+
+      technology = unpack_technology(opts)
+
+      task_uuid =
+        opts
+        |> Keyword.fetch!(:task_uuid)
+        |> Uptight.Base.mk_url!()
+
+      uuid = Commanded.UUID.uuid4() |> Uptight.Base.mk_url!()
+
+      solution_path = check_solution_path(opts)
+
+      %__MODULE__{
+        uuid: uuid,
+        hashed_identity: hashed_identity,
+        technology: technology,
+        task_uuid: task_uuid,
+        solution_path: solution_path,
+        submission_identity:
+          SubmissionIdentity.new(hashed_identity: hashed_identity, technology: technology)
+      }
+    end)
+  end
+
+  defp build_error(%{exception: %MatchError{term: %Result.Err{} = exception}}),
+    do: exception |> extract |> build_error()
+
+  defp build_error(%{exception: exception}), do: exception
+  defp build_error(exception), do: exception
+
+  defp check_solution_path(opts) do
+    solution_path = Keyword.fetch!(opts, :solution_path)
+
+    unless File.exists?(solution_path) do
+      raise ArgumentError, "Solution path is invalid: #{solution_path}"
+    end
+
+    T.new!(solution_path)
+  end
+
+  defp unpack_technology(opts) do
+    passed_technology_downcase =
+      opts
+      |> Keyword.fetch!(:technology)
+      |> String.downcase()
+
+    assert passed_technology_downcase in @supported_technologies,
+           "Technology #{passed_technology_downcase} is not supported"
```

Comment:
I think we write positive statements because when assertion fails the exception is "AssertionFailure"?..